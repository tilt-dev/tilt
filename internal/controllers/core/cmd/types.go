package cmd

import "github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"

type Cmd = v1alpha1.Cmd
type CmdSpec = v1alpha1.CmdSpec
type CmdStatus = v1alpha1.CmdStatus
type CmdStateRunning = v1alpha1.CmdStateRunning
type CmdStateTerminated = v1alpha1.CmdStateTerminated

const LabelManifest = v1alpha1.LabelManifest
