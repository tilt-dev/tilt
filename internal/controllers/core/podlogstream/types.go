package podlogstream

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type PodLogStream = v1alpha1.PodLogStream
type PodLogStreamSpec = v1alpha1.PodLogStreamSpec
type PodLogStreamStatus = v1alpha1.PodLogStreamStatus
type ContainerLogStreamStatus = v1alpha1.ContainerLogStreamStatus
type ObjectMeta = metav1.ObjectMeta
