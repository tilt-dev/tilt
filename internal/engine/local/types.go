package local

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type Cmd = v1alpha1.Cmd
type CmdList = v1alpha1.CmdList
type CmdStatus = v1alpha1.CmdStatus
type CmdSpec = v1alpha1.CmdSpec
type CmdStateWaiting = v1alpha1.CmdStateWaiting
type CmdStateTerminated = v1alpha1.CmdStateTerminated
type CmdStateRunning = v1alpha1.CmdStateRunning
type ObjectMeta = metav1.ObjectMeta
