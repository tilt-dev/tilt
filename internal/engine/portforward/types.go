package portforward

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type PortForward = v1alpha1.PortForward
type PortForwardSpec = v1alpha1.PortForwardSpec
type PortForwardStatus = v1alpha1.PortForwardStatus
type ObjectMeta = metav1.ObjectMeta
type Forward = v1alpha1.Forward
