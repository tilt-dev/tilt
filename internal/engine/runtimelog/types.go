package runtimelog

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type PodLogStream = v1alpha1.PodLogStream
type PodLogStreamSpec = v1alpha1.PodLogStreamSpec
type ObjectMeta = metav1.ObjectMeta
