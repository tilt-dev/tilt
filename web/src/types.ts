import { Alert } from "./alerts"
import { Facet } from "./facets"

export enum SocketState {
  Loading,
  Reconnecting,
  Closed,
  Active,
}

export enum ResourceView {
  Log,
  Alerts,
  Facets = 2,
}

export enum TriggerMode {
  TriggerModeAuto,
  TriggerModeManualAfterInitial,
  TriggerModeManualIncludingInitial,
}

// what is the status of the resource in the cluster
export enum RuntimeStatus {
  Ok = "ok",
  Pending = "pending",
  Error = "error",
  NotApplicable = "not_applicable",
}

// What is the status of the resource with respect to Tilt
export enum ResourceStatus {
  Building, // Tilt is actively doing something (e.g., docker build or kubectl apply)
  Pending, // not building, healthy, or unhealthy, but presumably on its way to one of those (e.g., queued to build, or ContainerCreating)
  Healthy, // e.g., build succeeded and pod is running and healthy
  Unhealthy, // e.g., last build failed, or CrashLoopBackOff
  Warning, // e.g., an undismissed restart
  None, // e.g., a manual build that has never executed
}

export type SnapshotHighlight = {
  beginningLogID: string
  endingLogID: string
  text: string
}

export enum ShowFatalErrorModal {
  Default,
  Show,
  Hide,
}

export type Snapshot = {
  view: Proto.webviewView
  isSidebarClosed: boolean
  path?: string
  snapshotHighlight?: SnapshotHighlight | null
}

// A plaintext representation of a line of the log,
// with metadata to render it in isolation.
//
// The metadata should be stored as primitive fields
// so that React's default caching behavior will kick in.
export type LogLine = {
  // We assume that 'text' does not contain a newline
  text: string
  manifestName: string
  level: string
}
