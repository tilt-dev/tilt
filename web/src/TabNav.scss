@import "constants";

.TabNav ul {
  list-style: none;
  display: flex;
}

.TabNav ul li:last-child {
  border-right: 1px solid $color-gray;
}

.TabNav ul li + li {
  border-left: 1px solid $color-gray;
}

.tabLink {
  box-sizing: border-box;
  height: $tabnav-height;
  flex: 0 0 auto;
  min-width: 10em;
  color: inherit;
  text-decoration: none;
  display: flex;
  align-items: center;
  justify-content: center;
  transition-property: color, background-color;
  transition-duration: $animation-timing;
  transition-timing-function: ease;
  @include button-text;
}

.tabLink:hover {
  color: $color-blue-light;
}

.tabLink--is-selected {
  right: 0;
  background-color: $color-gray-dark;
  color: $color-white;
  border-bottom: none;
}

.tabLink-alertBadge {
  font-family: $font-sans-serif;
  font-size: $font-size-smallest;
  background-color: $color-white;
  color: $color-gray-darkest;
  width: $alert-badge;
  height: $alert-badge;
  border-radius: $alert-badge;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: $spacing-unit / 3;
}
