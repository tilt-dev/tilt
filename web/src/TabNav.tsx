import React, { PureComponent } from "react"
import { Link } from "react-router-dom"
import { ResourceView } from "./types"
import "./TabNav.scss"

type NavProps = {
  previewUrl: string
  logUrl: string
  alertsUrl: string
  resourceView: ResourceView
  numberOfAlerts: number
}

class TabNav extends PureComponent<NavProps> {
  render() {
    let logIsSelected = this.props.resourceView === ResourceView.Log
    let previewIsSelected = this.props.resourceView === ResourceView.Preview
    let alertsIsSelected = this.props.resourceView === ResourceView.Alerts

    // The number of alerts should be for the selected resource
    return (
      <nav className="TabNav">
        <ul>
          <li>
            <Link
              className={`tabLink ${
                logIsSelected ? "tabLink--is-selected" : ""
              }`}
              to={this.props.logUrl}
            >
              Logs
            </Link>
          </li>
          <li>
            <Link
              className={`tabLink ${
                previewIsSelected ? "tabLink--is-selected" : ""
              }`}
              to={this.props.previewUrl}
            >
              Preview
            </Link>
          </li>
          <li>
            <Link
              className={`tabLink tabLink--errors ${
                alertsIsSelected ? "tabLink--is-selected" : ""
              }`}
              to={this.props.alertsUrl}
            >
              Alerts
              {this.props.numberOfAlerts > 0 ? (
                <span className="tabLink-alertBadge">
                  {this.props.numberOfAlerts}
                </span>
              ) : (
                ""
              )}
            </Link>
          </li>
        </ul>
      </nav>
    )
  }
}

export default TabNav
