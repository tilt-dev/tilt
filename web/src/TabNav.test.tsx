import { mount } from "enzyme"
import { createMemoryHistory, MemoryHistory } from "history"
import React from "react"
import { act } from "react-dom/test-utils"
import { Router } from "react-router"
import { OverviewNavProvider, TabNavContextConsumer } from "./TabNav"
import { ResourceName } from "./types"

// A fixture to help test the context provider
class Fixture {
  initialTabs: string[]
  nav: any = null
  root: any = null
  history: MemoryHistory = createMemoryHistory()
  validateTab: (res: string) => boolean = () => true

  constructor(initialTabs: string[]) {
    this.initialTabs = initialTabs
  }

  mount() {
    let tabs = this.nav?.tabs || this.initialTabs
    this.root = mount(
      <Router history={this.history}>
        <OverviewNavProvider
          tabsForTesting={tabs}
          validateTab={this.validateTab}
        >
          <TabNavContextConsumer>
            {(capturedNav) => void (this.nav = capturedNav)}
          </TabNavContextConsumer>
        </OverviewNavProvider>
      </Router>
    )
  }

  openResource(name: string, options?: any) {
    act(() => this.nav.openResource(name, options))

    // Enzyme doesn't properly re-render context providers with hooks,
    // so we manually re-render.
    this.mount()
  }
  closeTab(name: string) {
    act(() => this.nav.closeTab(name))

    // Enzyme doesn't properly re-render context providers with hooks,
    // so we manually re-render.
    this.mount()
  }
}

function newFixture(initialTabs: string[]): Fixture {
  let f = new Fixture(initialTabs)
  f.mount()
  return f
}

describe("tabnav", () => {
  afterEach(() => {
    localStorage.clear()
  })

  it("navigates to existing tab on click resource", () => {
    let f = newFixture(["res1", "res2"])
    expect(f.nav.tabs).toEqual(["res1", "res2"])
    expect(f.nav.selectedTab).toEqual("")

    f.openResource("res1")

    expect(f.nav.tabs).toEqual(["res1", "res2"])
    expect(f.nav.selectedTab).toEqual("res1")
    expect(f.history.location.pathname).toEqual("/r/res1/overview")
  })

  it("navigates to new tab on click resource", () => {
    let f = newFixture(["res1", "res2"])
    expect(f.nav.tabs).toEqual(["res1", "res2"])
    expect(f.nav.selectedTab).toEqual("")

    f.openResource("res3")

    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res3")
  })

  it("changes selected tab on click existing resource", () => {
    let f = newFixture(["res1", "res2", "res3"])
    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("")

    f.openResource("res1")

    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res1")

    f.openResource("res3")
    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res3")
  })

  it("changes selected tab on click new resource", () => {
    let f = newFixture(["res1", "res2", "res3"])
    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("")

    f.openResource("res2")

    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res2")

    f.openResource("res4")
    expect(f.nav.tabs).toEqual(["res1", "res4", "res3"])
    expect(f.nav.selectedTab).toEqual("res4")
  })

  it("open new tab to the right on double-click existing resource", () => {
    let f = newFixture(["res1", "res2", "res3"])
    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("")

    f.openResource("res1")

    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res1")

    f.openResource("res3", { newTab: true })
    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res3")
  })

  it("open new tab to the right on double-click new resource", () => {
    let f = newFixture(["res1", "res2", "res3"])
    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("")

    f.openResource("res1")

    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res1")

    f.openResource("res4", { newTab: true })
    expect(f.nav.tabs).toEqual(["res1", "res4", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res4")
  })

  it("navigates to the tab on the right when closing", () => {
    let f = newFixture(["res1", "res2", "res3"])
    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("")

    f.openResource("res2")

    expect(f.nav.tabs).toEqual(["res1", "res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res2")

    f.closeTab("res2")
    expect(f.nav.tabs).toEqual(["res1", "res3"])
    expect(f.nav.selectedTab).toEqual("res3")
    expect(f.history.location.pathname).toEqual("/r/res3/overview")
  })

  it("navigates to home on closing last tab when closing", () => {
    let f = newFixture(["res1"])
    expect(f.nav.tabs).toEqual(["res1"])
    expect(f.nav.selectedTab).toEqual("")

    f.openResource("res1")

    expect(f.nav.tabs).toEqual(["res1"])
    expect(f.nav.selectedTab).toEqual("res1")

    f.closeTab("res1")
    expect(f.nav.tabs).toEqual([ResourceName.all])
    expect(f.nav.selectedTab).toEqual("")
    expect(f.history.location.pathname).toEqual("/overview")
  })

  it("filters tabs that don't validate", () => {
    let f = new Fixture(["res1", "res2"])
    f.validateTab = (res) => res != "res1"
    f.history.location.pathname = "/r/res3/overview"
    f.mount()
    expect(f.nav.tabs).toEqual(["res2", "res3"])
    expect(f.nav.selectedTab).toEqual("res3")
  })

  it("filtering all tabs leaves the all tab", () => {
    let f = new Fixture(["res2", "res3"])
    f.validateTab = (res) => res == "res1"
    f.history.location.pathname = "/"
    f.mount()
    expect(f.nav.tabs).toEqual([ResourceName.all])
    expect(f.nav.selectedTab).toEqual("")
  })
})
