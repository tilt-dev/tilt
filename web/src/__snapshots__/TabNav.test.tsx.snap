// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`previews resources 1`] = `
<nav
  className="TabNav"
>
  <ul>
    <li>
      <a
        className="tabLink "
        href="/r/foo"
        onClick={[Function]}
      >
        Logs
      </a>
    </li>
    <li>
      <a
        className="tabLink tabLink--is-selected"
        href="/r/foo/preview"
        onClick={[Function]}
      >
        Preview
      </a>
    </li>
    <li>
      <a
        className="tabLink tabLink--errors "
        href="/r/foo/alerts"
        onClick={[Function]}
      >
        Alerts
        
      </a>
    </li>
  </ul>
</nav>
`;

exports[`shows error pane 1`] = `
<nav
  className="TabNav"
>
  <ul>
    <li>
      <a
        className="tabLink "
        href="/r/foo"
        onClick={[Function]}
      >
        Logs
      </a>
    </li>
    <li>
      <a
        className="tabLink "
        href="/r/foo/preview"
        onClick={[Function]}
      >
        Preview
      </a>
    </li>
    <li>
      <a
        className="tabLink tabLink--errors tabLink--is-selected"
        href="/r/foo/alerts"
        onClick={[Function]}
      >
        Alerts
        
      </a>
    </li>
  </ul>
</nav>
`;

exports[`shows logs 1`] = `
<nav
  className="TabNav"
>
  <ul>
    <li>
      <a
        className="tabLink tabLink--is-selected"
        href="/r/foo"
        onClick={[Function]}
      >
        Logs
      </a>
    </li>
    <li>
      <a
        className="tabLink "
        href="/r/foo/preview"
        onClick={[Function]}
      >
        Preview
      </a>
    </li>
    <li>
      <a
        className="tabLink tabLink--errors "
        href="/r/foo/alerts"
        onClick={[Function]}
      >
        Alerts
        
      </a>
    </li>
  </ul>
</nav>
`;

exports[`shows the number of errors in the error tab 1`] = `
<nav
  className="TabNav"
>
  <ul>
    <li>
      <a
        className="tabLink "
        href="/r/foo"
        onClick={[Function]}
      >
        Logs
      </a>
    </li>
    <li>
      <a
        className="tabLink "
        href="/r/foo/preview"
        onClick={[Function]}
      >
        Preview
      </a>
    </li>
    <li>
      <a
        className="tabLink tabLink--errors tabLink--is-selected"
        href="/r/foo/alerts"
        onClick={[Function]}
      >
        Alerts
        <span
          className="tabLink-alertBadge"
        >
          27
        </span>
      </a>
    </li>
  </ul>
</nav>
`;
